File: Exercises/Exercise 1 (Andreas)/shared variable/go/foo.go
// Use `go run foo.go` to run your program

package main

import (
    . "fmt"
    "runtime"
    "sync"
    "time"
)

var (
    i     = 0
    mutex sync.Mutex
)

func incrementing() {
    for j := 0; j < 1000001; j++ {
        mutex.Lock()
        i++
        mutex.Unlock()
    }
}

func decrementing() {
    for j := 0; j < 1000000; j++ {
        mutex.Lock()
        i--
        mutex.Unlock()
    }
}

func main() {
    runtime.GOMAXPROCS(2)

    go incrementing()
    go decrementing()

    time.Sleep(500 * time.Millisecond)
    Println("The magic number is:", i)
}

File: Exercises/Exercise 1 (Andreas)/shared variable/c/foo.c
// Compile with `gcc foo.c -Wall -std=gnu99 -lpthread`, or use the makefile
// The executable will be named `foo` if you use the makefile, or `a.out` if you use gcc directly

#include <pthread.h>
#include <stdio.h>

int i = 0;
pthread_mutex_t mutex;

void* incrementing() {

    for (int j = 0; j < 1000001; j++) {
        pthread_mutex_lock(&mutex);
        i++;
        pthread_mutex_unlock(&mutex);
    }

    return NULL;
}

void* decrementing() {

    for (int j = 0; j < 1000000; j++) {
        pthread_mutex_lock(&mutex);
        i--;
        pthread_mutex_unlock(&mutex);
    }

    return NULL;
}

int main() {
    pthread_t incrementingThread, decrementingThread;

    pthread_mutex_init(&mutex, NULL);

    pthread_create(&incrementingThread, NULL, incrementing, NULL);
    pthread_create(&decrementingThread, NULL, decrementing, NULL);

    pthread_join(incrementingThread, NULL);
    pthread_join(decrementingThread, NULL);

    printf("The magic number is: %d\n", i);
    return 0;
}